### PR TITLE
Add --force support to backendServiceSyncer.

### PR DESCRIPTION
**Pull request by G-Harmon**
We only want to modify an existing BackendService if the user supplies the --force flag.

Note: I still need to debug the healthcheck comparison handling, before I can properly (manually) test this.

Reviewers: @nikhiljindal @csbell @madhusudancs

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/27)

### CODE REVIEW

**Comment by G-Harmon** [comment]
I figured out why the healthcheck was always saying it needed to be updated. (The problem was that the 'desired' one we were generating didn't exact match what we get back from GCP.) I've added a 2nd commit to this PR to address it.
Once the health check comparison was working, I found that I needed to do something similar for the backend services- i.e. add more default values in to the 'desired' backend service.

**Comment by madhusudancs** [comment]
Reviewed 6 of 6 files at r1, 1 of 2 files at r2.
Review status: 6 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[app/kubemci/cmd/create.go, line 221 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoXCpeJCQk3jxFiREk:-KxoXCpeJCQk3jxFiREl:b4x9hls) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/cmd/create.go#L221)):*
> ```Go
> 			}
> 		}
> 		fmt.Println("Created cluster Ingress.")
> ```

Do we really need to print both? One of them is appropriate as a log with higher verbosity level right, perhaps "created" ones at V(4)??

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 152 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoYPvJKtOIeqKdf9ep:-KxoYPvJKtOIeqKdf9eq:b-3sa9ob) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L152)):*
> ```Go
> 	err := b.bsp.UpdateGlobalBackendService(desiredBE)
> 	if err != nil {
> 		fmt.Println("Error from UpdateGlobalBackendService", err)
> ```

Errors should be written to STDERR, no? But it's Ok for now if you can add a TODO.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 173 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoYfrR7z_-b1EyLDls:-KxoYfrR7z_-b1EyLDlt:b-dpyn44) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L173)):*
> ```Go
> 
> func backendServiceMatches(desiredBE, existingBE *compute.BackendService) bool {
> 	if desiredBE.AffinityCookieTtlSec != existingBE.AffinityCookieTtlSec ||
> ```

I guess you could negate all the comparisons and drop `if`, `else` and the returns. Something like:

```go
func backendServiceMatches(desiredBE, existingBE *compute.BackendService) bool {
	desiredBE.AffinityCookieTtlSec = existingBE.AffinityCookieTtlSec &&
        reflect.DeepEqual(desiredBE.Backends, existingBE.Backends) &&
        reflect.DeepEqual(desiredBE.CdnPolicy, existingBE.CdnPolicy) &&
        ...
```

It will be more readable.

---

*[app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go, line 122 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoZeSoHRfB8iK9sXpf:-KxoZeSoHRfB8iK9sXpg:boli7si) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/afe6941568729eff260cf2cfc06c9ea1055c6813/app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go#L122)):*
> ```Go
> 	if err == nil {
> 		fmt.Println("Health check", name, "exists already. Checking if it matches our desired health check")
> 		jsonExisting, _ := json.Marshal(existingHC)
> ```

Why this marshaling? Just for formatting? If so, just use `%#v` to format in `Infof` below?

Also, please don't silently drop errors. It might be borderline Ok here because you are using it just for formatting, but generally not recommended.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27)*

**Comment by G-Harmon** [comment]
Review status: 4 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[app/kubemci/cmd/create.go, line 221 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoXCpeJCQk3jxFiREk:-KxoqQvRkyF3-hqgsatd:b-k9ht47) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/cmd/create.go#L221)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Do we really need to print both? One of them is appropriate as a log with higher verbosity level right, perhaps "created" ones at V(4)??
</blockquote></details>

Done, though kept the "created" one on STDOUT. The problem I was trying to fix is that it looks like it does tons of work for the last cluster and almost nothing for the preceding ones. I want it to be clearer that it finished each ingress okay.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 152 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoYPvJKtOIeqKdf9ep:-KxonYh8i0Xi9AAi9XE5:bidvnp5) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L152)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Errors should be written to STDERR, no? But it's Ok for now if you can add a TODO.
</blockquote></details>

I don't think we've been trying for that.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 173 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoYfrR7z_-b1EyLDls:-KxooaGIQAFycNPtOG23:b4gs207) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L173)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I guess you could negate all the comparisons and drop `if`, `else` and the returns. Something like:

```go
func backendServiceMatches(desiredBE, existingBE *compute.BackendService) bool {
	desiredBE.AffinityCookieTtlSec = existingBE.AffinityCookieTtlSec &&
        reflect.DeepEqual(desiredBE.Backends, existingBE.Backends) &&
        reflect.DeepEqual(desiredBE.CdnPolicy, existingBE.CdnPolicy) &&
        ...
```

It will be more readable.
</blockquote></details>

sure, done. And this caught a bug! (s/existingBE/existingBE.Iap/ was the fix)

---

*[app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go, line 122 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoZeSoHRfB8iK9sXpf:-KxoooJgLPOeD5bCyF94:b-vwh32u) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/afe6941568729eff260cf2cfc06c9ea1055c6813/app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go#L122)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Why this marshaling? Just for formatting? If so, just use `%#v` to format in `Infof` below?

Also, please don't silently drop errors. It might be borderline Ok here because you are using it just for formatting, but generally not recommended.
</blockquote></details>

I tried %#v, but I still didn't see inside nested structs, which was important.
Agree that we shouldn't usually drop errors. This was only for logging... anyway added an Infof with the error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27)*

**Comment by madhusudancs** [comment]
Reviewed 2 of 3 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[app/kubemci/cmd/create.go, line 221 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoXCpeJCQk3jxFiREk:-Kxp1QdamTP_gGwLdDxn:b-pu4g1r) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/cmd/create.go#L221)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Done, though kept the "created" one on STDOUT. The problem I was trying to fix is that it looks like it does tons of work for the last cluster and almost nothing for the preceding ones. I want it to be clearer that it finished each ingress okay.
</blockquote></details>

Ok fair enough.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 152 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoYPvJKtOIeqKdf9ep:-Kxp1kehK6nUCuwLbp-F:b-xpqcx1) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L152)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

I don't think we've been trying for that.
</blockquote></details>

Fair enough. That's how it looks like :)

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 173 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoYfrR7z_-b1EyLDls:-Kxp33AMu42t4Cy7_g2m:b-83sjoh) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/2ad14ce74aab42b3bbdeca2943fbc822fd78c8ec/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L173)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

sure, done. And this caught a bug! (s/existingBE/existingBE.Iap/ was the fix)
</blockquote></details>

Nice!

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 125 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-Kxp2yTedjrGvCi5GIxo:-Kxp2yTedjrGvCi5GIxp:b-75y304) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L125)):*
> ```Go
> 		jsonDesired, mErr := json.Marshal(desiredBE)
> 		if mErr != nil {
> 			glog.Infof("Error marshalling dsired backend: %v", mErr)
> ```

s/dsired/desired/

---

*[app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go, line 122 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxoZeSoHRfB8iK9sXpf:-Kxp3NOZuQjlQuZzBWEq:bsjv2by) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/afe6941568729eff260cf2cfc06c9ea1055c6813/app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go#L122)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

I tried %#v, but I still didn't see inside nested structs, which was important.
Agree that we shouldn't usually drop errors. This was only for logging... anyway added an Infof with the error.
</blockquote></details>

Ok for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27)*

File: app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go [hunk 1]
```diff
@@ -41,43 +42,87 @@ func TestEnsureBackendService(t *testing.T) {
...
+		}
+		if be.Port != port || be.PortName != portName {
+			t.Errorf("unexpected port and port name, expected: %s/%s, got: %s/%s", portName, port, be.PortName, be.Port)
+		}
+		if len(be.Backends) != 1 || be.Backends[0].Group != igLink {
+			t.Errorf("unexpected backends in backend service. expected one backend for %s, got: %v", igLink, be.Backends)
+		}
 	}
+
 	// TODO(nikhiljindal): Test update existing backend service.
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go]: Remove this TODO now :)

File: app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go [hunk 2]
```diff
@@ -174,6 +211,12 @@ func (b *BackendServiceSyncer) desiredBackendService(lbName string, port ingress
 		Port:         port.Port,
 		PortName:     portName,
 		Backends:     desiredBackends(igLinks),
+		// We have to fill in these fields so we can properly compare to what's returned to us
+		ConnectionDraining:  &compute.ConnectionDraining{},
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go]: We should get these values from existing if it exists and skip them otherwise, rather than hardcoding them here.
Please add a TODO to fix this

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go [hunk 3]
```diff
@@ -201,13 +204,14 @@ func (h *HealthCheckSyncer) desiredHealthCheck(lbName string, port ingressbe.Ser
...
 		Type:               string(port.Protocol),
-		// TODO: Try Kind: compute#healthCheck
+		Kind:               "compute#healthCheck",
 	}
 	switch port.Protocol {
 	case "HTTP":
 		hc.HttpHealthCheck = &compute.HTTPHealthCheck{
 			Port:        port.Port,
 			RequestPath: "/healthz", // TODO(nikhiljindal): Allow customization.
+			ProxyHeader: "NONE",
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: I assume we will require this for HTTPs as well? Fine to do it in another PR, since HTTPs doesnt work right now anyway.

**Comment by nikhiljindal** [comment]
thx lgtm. One minor comment to add a TODO and another to remove a TODO :)

File: app/kubemci/cmd/create.go [hunk 4]
```diff
@@ -197,7 +197,7 @@ func createIngressInClusters(kubeconfig, ingressFilename string, clusters []stri
 	// TODO(nikhiljindal): Validate and optionally add the gce-multi-cluster class annotation to the ingress YAML spec.
 	var err error
 	for _, c := range clusters {
-		fmt.Println("\nHandling context:", c)
+		glog.V(4).Infof("Creating Ingress for cluster: %v...", c)
```
Comment (perotinus) [inline app/kubemci/cmd/create.go]: nit: "in cluster"

File: app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go [hunk 5]
```diff
@@ -114,16 +116,34 @@ func (b *BackendServiceSyncer) ensureBackendService(lbName string, port ingressb
...
-		fmt.Println("Updating existing backend service", name, "to match the desired state")
-		return b.updateBackendService(desiredBE)
+		if forceUpdate {
+			// TODO(G-Harmon): Figure out how to properly calculate the fp. Using Sha256 returned a googleapi error.
+			desiredBE.Fingerprint = existingBE.Fingerprint
+			fmt.Println("Updating existing backend service", name, "to match the desired state (since --force specified)")
+			return b.updateBackendService(desiredBE)
+		} else {
+			// TODO(G-Harmon): Show diff to user and prompt yes/no for overwriting.
+			fmt.Println("Will not overwrite a differing BackendService without the --force flag.")
```
Comment (perotinus) [inline app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go]: Could you extract the logic above that shows JSON versions into a method and call that here too?

(not necessary now)

File: app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go [hunk 6]
```diff
@@ -136,9 +156,10 @@ func (b *BackendServiceSyncer) ensureBackendService(lbName string, port ingressb
 // updateBackendService updates the backend service and returns the updated backend service.
 func (b *BackendServiceSyncer) updateBackendService(desiredBE *compute.BackendService) (*compute.BackendService, error) {
 	name := desiredBE.Name
-	fmt.Println("Updating existing backend service", name, "to match the desired state")
 	err := b.bsp.UpdateGlobalBackendService(desiredBE)
 	if err != nil {
+		// TODO(G-Harmon): Errors should probably go to STDERR.
+		fmt.Println("Error from UpdateGlobalBackendService", err)
```
Comment (perotinus) [inline app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go]: You probably want `Printf` here.

File: app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go [hunk 7]
```diff
@@ -159,9 +180,25 @@ func (b *BackendServiceSyncer) createBackendService(desiredBE *compute.BackendSe
...
-	// TODO(nikhiljindal): Add proper logic to figure out if the 2 backend services match.
-	// Need to add the --force flag for user to consent overritting before this method can be updated to return false.
-	return true
+	return desiredBE.AffinityCookieTtlSec == existingBE.AffinityCookieTtlSec &&
+		reflect.DeepEqual(desiredBE.Backends, existingBE.Backends) &&
+		reflect.DeepEqual(desiredBE.CdnPolicy, existingBE.CdnPolicy) &&
+		reflect.DeepEqual(desiredBE.ConnectionDraining, existingBE.ConnectionDraining) &&
+		desiredBE.Description == existingBE.Description &&
+		desiredBE.EnableCDN == existingBE.EnableCDN &&
+		// Check fingerprint? fp is not required for insert but is required for update.
```
Comment (perotinus) [inline app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go]: Is this a TODO?

File: app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go [hunk 8]
```diff
@@ -109,7 +154,7 @@ func TestDeleteBackendService(t *testing.T) {
 			Port: port,
 			Name: portName,
 		},
-	}, []string{igLink}); err != nil {
+	}, []string{igLink}, false /*forceUpdate*/); err != nil {
```
Comment (perotinus) [inline app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go]: Should the `/*forceUpdate*/` be removed?

Comment (madhusudancs) [reply app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go]: Probably not. It is poor man's keyword argument.

Although, I must admit that I find this form of complex object construction directly in a function call statement absurd.

Comment (nikhiljindal) [reply app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go]: +1 to keeping it as documentation

**Comment by nikhiljindal** [comment]
@G-Harmon needs rebase

**Comment by G-Harmon** [comment]
rebased.

---

Reviewed 2 of 6 files at r1, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[app/kubemci/cmd/create.go, line 200 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxnbSaEpIVgNzBLfb38-r3-200:-Kxq-R4MwCu94ukTqCR5:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/cmd/create.go#L200)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

nit: "in cluster"
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 125 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-Kxp2yTedjrGvCi5GIxo:-KxpvbbQIlGUHhCzK7ke:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L125)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/dsired/desired/
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 142 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxnbSaEpIVgNzBLfb39-r3-142:-KxpwdBFYulOy7dIPKNu:b-f5tng0) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L142)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Could you extract the logic above that shows JSON versions into a method and call that here too?

(not necessary now)
</blockquote></details>

Is there a good way to actually print just the diff? That would be much more useful than having to compare these things manually. (and when there are fields we ignore.)
Added a TODO

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 162 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxnbSaEpIVgNzBLfb39-r3-162:-KxpybbUu-eTuSIERk6M:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L162)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

You probably want `Printf` here.
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 189 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxnbSaEpIVgNzBLfb39-r3-189:-KxpyvJuwRMwHJeqQxDS:b-b7oleq) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L189)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Is this a TODO?
</blockquote></details>

Yes, added 'todo'.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go, line 215 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxnbSaEpIVgNzBLfb39-r3-215:-KxpzTPwAnYY-9Jj7d5A:b-md4frx) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer.go#L215)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

We should get these values from existing if it exists and skip them otherwise, rather than hardcoding them here.
Please add a TODO to fix this
</blockquote></details>

TODO added.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go, line 126 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxnbSaEpIVgNzBLfb3A-r3-126:-KxpziL7QQyoyxATMy33:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go#L126)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Remove this TODO now :)
</blockquote></details>

Done.

---

*[app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go, line 157 at r3](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-KxnbSaEpIVgNzBLfb3A-r3-157:-KxpzsivXQPfcaMlIOtV:beq07nu) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/ba7c0f4c50ea2d3e1ec2edbd1baa27d6225f0be6/app/kubemci/pkg/gcp/backendservice/backendservicesyncer_test.go#L157)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

+1 to keeping it as documentation
</blockquote></details>

agree w/ Madhu that this is a crazy thing you can do in Go (and kind of weird).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/27#-:-KxqH0Jfn1bm5O3ThWEh:bmaduqb)*

**Comment by nikhiljindal** [comment]
@madhusudancs @perotinus any last comments before we merge this?

**Comment by madhusudancs** [comment]
@nikhiljindal it's all good now.

/lgtm
